tao_compiler/mlir/disc/utils/source_emitter.cc

```diff
@@ -0,0 +1,601 @@
+  } else if (isa<lmhlo::NotOp>(op)) {
+    result_name = EmitUniqueName("not");
+    expression =
+        CUDAMathFuncName<lmhlo::NotOp>(result_type) + "(" + input_str + ")";
+  } else if (isa<lmhlo::RsqrtOp>(op)) {
+    result_name = EmitUniqueName("rsqrt");
+    expression =
+        CUDAMathFuncName<lmhlo::RsqrtOp>(result_type) + "(" + input_str + ")";
+  } else if (isa<lmhlo::SqrtOp>(op)) {
+    result_name = EmitUniqueName("sqrt");
+    expression =
+        CUDAMathFuncName<lmhlo::SqrtOp>(result_type) + "(" + input_str + ")";
+  } else if (isa<lmhlo::TanhOp>(op)) {
+    result_name = EmitUniqueName("tanh");
+    expression =
+        CUDAMathFuncName<lmhlo::TanhOp>(result_type) + "(" + input_str + ")";
+  }
+
+  assert(binding.count(op->getOperand(1)) == 0);
+  binding[op->getOperand(1)] = result_name;
+
+  return type_str + " " + result_name + " = " + expression;
+}
+
+llvm::Optional<std::string> SourceEmitterCUDA::EmitElemWiseBinaryOp(
+    Operation* op, ValueNameBinding& binding) {
+  auto lhs = op->getOperand(0);
+  auto rhs = op->getOperand(1);
+  if (binding.count(lhs) == 0 || binding.count(rhs) == 0) {
+    llvm::None;
+  }
+  std::string lhs_str = binding[lhs];
+  std::string rhs_str = binding[rhs];
+
+  Type result_type =
+      op->getOperand(2).getType().cast<MemRefType>().getElementType();
+  std::string type_str = MLIRType2CUDATypeStr(result_type);
+
+  std::string result_name;
+  std::string expression;
+  if (isa<lmhlo::AddOp>(op)) {
+    result_name = EmitUniqueName("add");
+    expression = lhs_str + " + " + rhs_str;
+  } else if (isa<lmhlo::SubtractOp>(op)) {
+    result_name = EmitUniqueName("subtract");
+    expression = lhs_str + " - " + rhs_str;
+  } else if (isa<lmhlo::MulOp>(op)) {
+    result_name = EmitUniqueName("mul");
+    expression = lhs_str + " * " + rhs_str;
+  } else if (isa<lmhlo::DivOp>(op)) {
+    result_name = EmitUniqueName("div");
+    expression = lhs_str + " / " + rhs_str;
+  } else if (isa<lmhlo::MaxOp>(op)) {
+    result_name = EmitUniqueName("max");
+    expression = lhs_str + " > " + rhs_str + " ? " + lhs_str + " : " + rhs_str;
+  } else if (isa<lmhlo::MinOp>(op)) {
+    result_name = EmitUniqueName("min");
+    expression = lhs_str + " < " + rhs_str + " ? " + lhs_str + " : " + rhs_str;
+  } else if (auto compare = dyn_cast_or_null<lmhlo::CompareOp>(op)) {
+    result_name = EmitUniqueName("compare");
+    std::string cmp_str;
+    switch (compare.getComparisonDirection()) {
+      case mhlo::ComparisonDirection::EQ:
+        cmp_str = "==";
+        break;
+      case mhlo::ComparisonDirection::NE:
+        cmp_str = "!=";
+        break;
+      case mhlo::ComparisonDirection::LT:
+        cmp_str = "<";
+        break;
+      case mhlo::ComparisonDirection::LE:
+        cmp_str = "<=";
+        break;
+      case mhlo::ComparisonDirection::GT:
+        cmp_str = ">";
+        break;
+      case mhlo::ComparisonDirection::GE:
+        cmp_str = ">=";
+        break;
```

Comment:
nit: better to have a `default` branch to ensure exhausted enumeration.

Reply:
The switch has already enumerated all possible values of ComparisonDirection.

Reply:
Yes, but what if anyone do some change in the future, say, deleting a branch by mistake.  Don't trust anyone include yourself, let your code defend itself ~~~ (just joking~~)